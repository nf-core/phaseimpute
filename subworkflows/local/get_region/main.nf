workflow GET_REGION {
    take:
        input_region // Region string to use ["all", "chr1", "chr1:0-1000"]
        ch_fasta     // [[genome], fasta, fai]

    main:
        ch_versions      = Channel.empty()

        // Gather regions to use and create the meta map
        if (input_region ==~ '^(chr)?[0-9XYM]+$' || input_region == "all") {
            ch_regions = ch_fasta.map{it -> it[2]}
                .splitCsv(header: ["chr", "size", "offset", "lidebase", "linewidth", "qualoffset"], sep: "\t")
                .map{it -> [chr:it.chr, region:"0-"+it.size]}
            if (input_region != "all") {
                ch_regions = ch_regions.filter{it.chr == input_region}
            }
            ch_regions = ch_regions
                .map{ [[chr: it.chr, region: it.chr + ":" + it.region], it.chr + ":" + it.region]}
        } else {
            if (input_region ==~ '^chr[0-9XYM]+:[0-9]+-[0-9]+$') {
                ch_regions = Channel.from([input_region])
                    .map{ [[chr: it.split(":")[0], "region": it], it]}
            } else {
                error "Invalid input_region: ${input_region}"
            }
        }

    emit:
        regions           = ch_regions       // channel: [ [chr, region], region ]
        versions          = ch_versions      // channel: [ versions.yml ]
}
