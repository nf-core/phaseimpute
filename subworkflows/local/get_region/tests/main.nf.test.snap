{
    "Should run with specified region without fasta": {
        "content": [
            [
                [
                    {
                        "chr": "chr22",
                        "region": "chr22:0-4000"
                    },
                    "chr22:0-4000"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-17T15:20:16.458964814"
    },
    "Should run with an error": {
        "content": [
            [
                "N E X T F L O W  ~  version 23.10.1",
                "Launching `/mnt/c/Users/llenezet/Documents/Repositories/phaseimpute/.nf-test/tests/68cf57ba96afcf7b8ba9414fb03e77b1/meta/mock.nf` [angry_babbage] DSL2 - revision: d7db2c206c",
                "Invalid input_region: chr22:0-4000:4648"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T10:00:37.781505704"
    },
    "Should run with specified chr": {
        "content": [
            [
                [
                    {
                        "chr": "chr22",
                        "region": "chr22:0-50818468"
                    },
                    "chr22:0-50818468"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T09:59:50.83303021"
    },
    "Should run with 'all'": {
        "content": [
            [
                [
                    {
                        "chr": "chr21",
                        "region": "chr21:0-46709983"
                    },
                    "chr21:0-46709983"
                ],
                [
                    {
                        "chr": "chr22",
                        "region": "chr22:0-50818468"
                    },
                    "chr22:0-50818468"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T09:59:45.078764008"
    }
}